### ISSUE TITLE
[Feature Request] (BFIv3) Emulate a CRT Electron Gun Via Rolling-Scan BFI

### ISSUE THREAD

**Issue by mdrejhon** [libretro/RetroArch#10757]
# New Methods of Emulating CRT Via Sheer Hz

I'm the founder of Blur Busters and creator of TestUFO. Today, we now have 240Hz 1ms IPS panels, and DELL is releasing a 360Hz IPS monitor this summer. This is an amazing opportunity for emulation.

**Glossary**
BFI = **B**lack **F**rame **I**nsertion, used to reduce motion blur on LCDs to mimic a CRT
BFIv1 = Classic 60fps at 120Hz BFI, _already implemented in RetroArch_
BFIv2 = Improvements to BFI for higher Hz, _See GitHub #10754_
BFIv3 = Rolling-bar BFI, emulating CRT electron gun temporally, **FOCUS OF THIS GITHUB ITEM**

Inspired by a [forum thread on ArcadeControls](http://forum.arcadecontrols.com/index.php/topic,162926.0.html) as well as [MAME Temporal HLSL](https://github.com/mamedev/mame/issues/6762), I'm crossposting the request, with some modifications.

## Rolling-Bar Black Frame Insertion Concept

It can even achieve sub-refresh latencies using ordinary VSYNC ON.   One wouldn't have to care about how a display refreshes.  For example, at 240Hz, you'd display 1/4th of an emulator refresh cycle (rasterplotted in real time), in 4 separate output frames.  

**For a 60Hz emulator module onto a 240Hz monitor**
* Emulate 1/240sec, output 1/4-height frame at top
* Emulate 1/240sec, output 1/4-height frame just above center
* Emulate 1/240sec, output 1/4-height frame just below center
* Emulate 1/240sec, output 1/4-height frame at bottom

The rest of the frame would be black (except for any required alphablend overlaps to eliminate seams/artifacts)

Make top/bottom edges of bars fuzzy to reduce/eliminate tearing artifacts (avoid emulating the look of VSYNC OFF tearing). I have confirmed that you have to overlap the bars between output refresh cycles.  Use alpha-blend-to-black slightly beyond the 1/4-height frames for the 240Hz situation example.  This will prevent tearing artifacts. Make sure to gamma-correct the alphablend overlaps. Remember that RGB(128,128,128) is not exactly half the photons of RGB(255,255,255).  Use a configurable gamma correction number in HLSL config file.

## It can also acomplish beam raced latency without black frames too

The rolling scan BFI could in theory be configurable to full persistence.  It would achieve beam raced latencies to the output refresh cycle granularity.

**For a 60Hz emulator module onto a 240Hz monitor**
* Emulate 1/240sec, output frame with top 1/4 new refresh, bottom 3/4 old refresh
* Emulate 1/240sec, output frame with top 1/2 new refresh, bottom 1/2 old refresh
* Emulate 1/240sec, output frame with top 3/4 new refresh, bottom 1/4 old refresh
* Emulate 1/240sec, output frame with whole screen new refresh

Make sure to alphablend the seams (blur the refresh overlap line) to prevent tearing artifacts, otherwise it looks like "a VSYNC OFF emulation in VSYNC ON".  The alphablend fixes the tearing.  

This allows beam-raced latencies via sheer brute Hz.  So, this module could be configurable to have full persistence or rolling low persistence (rolling black period).

This will scale very well to the future, in the refresh rate race to retina refresh rates too.  A future 1000Hz display would emulate original machine latency to an error margin of 1ms, duplicating sub-refresh original-machine latencies, regardless of display technology (scan direction, display refresh pattern, etc).   Most high-Hz gaming displays have a sub-refresh processing delay, so the higher Hz you go, the more it converges into original-machine latency.  

Very little crossplatform dependencies, the only thing needed is VSYNC ON and the ability to do framerate=Hz.  

## Emulator Hz and destination Hz doesn't need to be divisible.

**Situation Example of 60Hz CRT emulation onto a 200Hz LCD** ... This formula is very Hz-agnostic.

Emulator Refresh Cycle 1:
....Real Refresh 1: full 60/200th height bar (30% screen height), at 0%-30% vertical position
....Real Refresh 2: full 60/200th height bar (30% screen height), at 30%-60% vertical position
....Real Refresh 3: fuill 60/200th height bar (30% screen height), at 60%-90% vertical position
....Real Refresh 4: 1/3 of 60/200th height bar (10% screen height), at 90%-100% vertical position

Emulator Refresh Cycle 2:
....Real Refresh 5: 2/3 of 60/200th height bar (20% screen height), at 0%-20% vertical position
....Real Refresh 6: full 60/200th height bar (30% screen height), at 20%-50% vertical position
....Real Refresh 7: full 60/200th height bar (30% screen height), at 50%-80% vertical position
....Real Refresh 8: 2/3 of of 60/200th height bar (20% screen height), at 80%-100% vertical position

Emulator Refresh Cycle 3:
....Real Refresh 9: 1/3 of 60/200th height bar (10% screen height), at 0%-10% vertical position
....Real Refresh 10: full 60/200th height bar (30% screen height), at 10%-40% vertical position
....Real Refresh 11: full 60/200th height bar (30% screen height), at 40%-70% vertical position
....Real Refresh 12: full 60/200th height bar (30% screen height), at 70%-100% vertical position

This is just a conceptual example of temporal compensation.  Algorithmically, this can be used for black frame insertion (rolling bar, some bars with image data, other bars black, with alphablended bleed overlap) or for "beam racing via brute Hz" (all bars with image data, new emu Hz overwriting old emu Hz, alphablend/blur the seams) or both simultaneously (rolling BFI + beam racing simultaneously).   You could adjust the alphablend factor up/down.

## Theoretically can also acomplish stutterless standards conversion (50Hz onto 60Hz)

Using a SIMILAR formula above, combined with a huge overlap alphablend adjustment (i.e. 50% or 75% screen height of overlap, perhaps).  Basically a scanout-enhanced version of a common alphablend standards-conversion algorithm.   It would thus, eliminate stutters.  The taller the alphablend overlap, the less stutter of standards-conversion.  The bar-overlap should be a configurable parameter in the configuration file.

This would conceptually be a more advanced version of the software-based variable refresh animation: www.testufo.com/vrr ....where I'm able to play any framerate stutterlessly onto any refreshrate, with a very simple alphablend algorithm not too dissimilar from the common 50/60 standards conversion alphablend algorithm.

So we're just "abusing" the alphablend overlap feature of this theoretical "Temporal HLSL" as the method of destutter in much the same stutter-eliminating way.

Apparently, @TomHarte already tested something similar (some kind of scanout-alphablend algorithm) in his experiments with CLK.

## Temporal HLSL concept is a universal CRT emulator with amazing crossplatform flexibility

It scale up and down universally

- Can temporally emulate a CRT electron gun at the granularity of destnation Hz
- Map any Hz to any Hz without stutter
- Benefits for stutterles Hz-conversion (50Hz onto 60Hz) 
- Benefits for high Hz (i.e. 60Hz onto 360Hz) 
- Potentially beam raceable (Temporal HLSL = a virtualized software display)
- Flickerless option (full persistence option)
- Flicker option (simulate CRT phosphor, including decay, subject to output-Hz granularity)

## OPTIONAL: Theoretically compatible with hardware beamraced VSYNC 

See #6984 for Beam Raced VSYNC

**<Optional>**
- Temporal HLSL could emulate a higher-Hz display internally independently of output Hz
- A flywheel sync algorithm can optionally kick in whenever emuHz = outputHz, allowing a separate module to frameslice-beamrace this virtualized internal display to a low-Hz output display.

Thus, it would produce latencies identical to the Lagless VSYNC / Beamraced VSYNC approaches, like the one already implemented in WinUAE https://github.com/tonioni/WinUAE/issues/133 

Certainly, actual-hardware beam racing would be 100% optional (since it requires minor platform-dependant and display-scanout-direction knowledge)

I only simply add this section to say that Temporal HLSL is 100% compatible with hardware-based beam racing simply by internally virtualizing a higher-Hz display internally, e.g. doing 600 frames per second of Temporal HLSL into RAM, and then doing 600-frameslice beamracing onto a real 60Hz display, subject to jitter margin to hide problems caused by CRT-curvature algorithms as well as computer performance jitter.

**Software-based beam racing**: Output Hz massively above Emulator Hz. The art of emulating a CRT electron gun via brute Hz.
**Hardware-based beam racing**: Output Hz same as Emulator Hz. Beam raced synchronization of rasterplotting emulator's raster ahead of real display's raster. Using beam raced VSYNC OFF frameslices, that would work off a Temporal HLSL framebuffer too. (An approach successfully already implemented in a few emulators)

Basically the Temporal HLSL concept is compatible with both software-based beam racing and hardware-based beam racing.
**</Optional>**

**Comment by mdrejhon** [comment]
GitHub item #10758 is a pre-requisite for this.

I broke out that specific pre-requisite task separately.

**Comment by mdrejhon** [comment]
# Frequently Asked Questions

## Question: How Much Display Motion Blur Can I Reduce?

# Answer 1: Answer For Sample-And-Hold Displays like LCDs without using Blur Reduction Mode

Software BFI also reduces motion blur without a strobe backlight, up to these following numbers:

120Hz refresh rate: **50%** motion blur reduction for software BFI for emulator 60Hz
180Hz refresh rate: **66%** motion blur reduction for software BFI for emulator 60Hz
240Hz refresh rate: **75%** motion blur reduction for software BFI for emulator 60Hz
300Hz refresh rate: **80%** motion blur reduction for software BFI for emulator 60Hz
360Hz refresh rate: **83%** motion blur reduction for software BFI for emulator 60Hz
480Hz refresh rate: **88%** motion blur reduction for software BFI for emulator 60Hz
1000Hz refresh rate: **94%** motion blur reduction for software BFI on emulator 60Hz

Successfully demo'd in TestUFO (when run on 240Hz+ monitors), the motion blur amount will be (Source-Hz) divided by (Destination-Hz) -- persistence refresh cycle visibility time.  The higher Hz, the easier to emulate shorter-persistence phoshpor.

Display motion blur is directly proportional to pixel visibility time, as seen in www.testufo.com/eyetracking -- We're simply using brute Hz for ON-then-OFF to simulate the briefness of phosphor.  The higher Hz, the briefer we can make pixels flash.  This is true regardless for BFIv2 (#10754) or BFIv3 approach (this github)

# Answer 2: For impulsed displays such as LightBoost / ULMB / ELMB / DyAc / VRB etc displays

Strobe backlights are the ones found in [Motion Blur Reduction FAQ](https://blurbusters.com/faq/motion-blur-reduction).

If a display has a strobe backlight mode, see #10754 for how software BFI simultaneously helps hardware strobing.  While software BFI with a hardware strobe backlight, will usually work better in a global mode.

## Q: Does a Rolling Scan BFI Flicker Less Than Global Strobing?

# A: Yes.

A rolling BFI keeps average number of photons hitting human eyeballs consistent (because the global flashing strobe of LightBoost / ULMB / etc).  So rolling BFI is more eye-friendly than global-flash BFI (only 60 global flashes per second, complete blackness in between).   

A CRT tube is usually always emitting light somewhere on the tube surface -- because of the scanning phosphor dot.  That softens the harshness of flicker a little, and explains why 60Hz global-strobe looks more flickery than 60Hz scanning-strobe.

**Comment by mdrejhon** [comment]
UPDATE: There is an item that may make BFIv3 easier/more flexible.

# [[Feature Request] Futureproof RetroArch with precision frame pacing presenter thread](https://github.com/libretro/RetroArch/issues/11390)

**Comment by mdrejhon** [comment]
## Interesting Equivalence: Conceptual Emulation of a High Speed Video of a CRT Tube

One conceptual way to more easily understand this github item:

A high speed video of a CRT tube.  You see a rolling bar in those, with blurry edges.  This is seen in many YouTube -- you see phosphor trailing behind.

**Did you know that playback of a 1000fps high-speed video of a CRT tube -- in real time onto a true 1000Hz display -- allows a 1000Hz display to perceptually emulate the original CRT?**

The "Temporal HLSL" concept, aims to emulate that behavior in software.  For best emulation perfection requires retina resolution (spatial HLSL) + retina refresh rate (temporal HLSL) + retina HDR (to keep it beam-emulation bright).   

A display that combines all of this will take some time to arrive, but -- this encompasses the venn diagram of capturing the look-feel of a CRT tube (at flat tubes).

(plus a slight amount of edge-alphablending for overlaps to prevent tearing artifacts from appearing, especially for the sharp bottom edge of the rolling-scan bar as seen in high speed videos.)

**Comment by mdrejhon** [comment]
BTW, 240Hz OLEDs actually make CRT electron beam simulators pretty workable today.  Some early custom internal TestUFO experiment shows that you only need 240Hz in order for CRT electron beam simulators to become superior to monolithic BFI, at least when it comes to instant-pixel-response displays such as OLED.   You get a fair bit of light loss, but you also get that with ordinary BFI and strobing too.

So beginning 2023, superior-than-BFI electron beam simulators is already viable!

Strobing reduces motion blur more at 240Hz (need 1000Hz for software-based CRT electron beam simulators to be as low-blur as hardware strobing), but the 60Hz flicker is a LOT less painful.   Since the rolling-scan means photons are hitting eyeballs at all times, and the simulated phosphor decay softens the flicker.

**Comment by mdrejhon** [comment]
I have a surprise coming December 24, 2024 -- the world's most accurate CRT electron beam simulator shader.

**Comment by mdrejhon** [comment]
The algorithm is ready!

I've published the article:
www.blurbusters.com/crt

I've released a MIT-licensed open source shader: 
www.github.com/blurbusters/crt-beam-simulator

Shadertoy animation demo (for 240Hz)
www.shadertoy.com/view/XfKfWd

![crt-simulation-animated](https://github.com/user-attachments/assets/0f00ece6-c71e-489b-8831-10fec148f682)

Can easily adjust settings for 120Hz or 144Hz or 165Hz or 360Hz or 480Hz or 540Hz!

Please implement it into Retroarch.  Pretty please.

**Comment by hizzlekizzle** [comment]
Hey, looks great. That's awesome that Mr. Lottes was able to help with it. I'm a big fan of his work.

I'll see about getting this ported over ASAP. Shouldn't be too tough.

**Comment by mdrejhon** [comment]
Fantastique. Please expose these adjustments:

- Brightness vs Clarity (GAIN_VS_BLUR)
- Gamma (GAMMA)

**Comment by mdrejhon** [comment]
Crossposted:

> Porting to RetroArch was a breeze. It's available here now: https://github.com/libretro/slang-shaders/blob/master/subframe-bfi/shaders/crt-beam-simulator.slang and will show up in the online updater in a few minutes/hours.
> I replaced the edge-blended version I had made with it, since this one is superior in every way lol.

That was damn fast! Nice Christmas surprise. And I can combine CRT filters simultaneously too? Neat!

You should rename the menu in RetroArch if possible, to at least catch attention -- that it is a new better shader. Also eventually, I'll add a phase-offset since I can reduce the latency of this CRT simulator by probably 1 frameslice (1 native refresh cycle) simply by adding +time as a constant. I need to experiment with my changes in ShaderToy in the coming week (It's Christmas).

But it's absurdly fantastic to see an actual deployment the same day I released my simulation shader!

Once it's shipping on PC/Mac/Linux/iPhone/iPad, we can close this.  Maybe mark this item as completed.
